montecarlo/mcpy/monte_carlo.py

```diff
@@ -0,0 +1,169 @@
+    assert 'dgps' in config, "config dict must contain dgps"
+    assert 'dgp_opts' in config, "config dict must contain dgp_opts"
+    assert 'method_opts' in config, "config dict must contain method_opts"
+    assert 'mc_opts' in config, "config dict must contain mc_opts"
+    assert 'metrics' in config, "config dict must contain metrics"
+    assert 'methods' in config, "config dict must contain methods"
+    assert 'plots' in config, "config dict must contain plots"
+    assert 'target_dir' in config, "config must contain target_dir"
+    assert 'reload_results' in config, "config must contain reload_results"
+    assert 'n_experiments' in config['mc_opts'], "config[mc_opts] must contain n_experiments"
+    assert 'seed' in config['mc_opts'], "config[mc_opts] must contain seed"
+
+class MonteCarlo:
+
+    def __init__(self, config):
+        self.config = config
+        _check_valid_config(self.config)
+        config['param_str'] = '_'.join(['{}_{}'.format(filesafe(k), v) for k,v in self.config['mc_opts'].items()])
+        config['param_str'] += '_' + '_'.join(['{}_{}'.format(filesafe(k), v) for k,v in self.config['dgp_opts'].items()])
+        config['param_str'] += '_' + '_'.join(['{}_{}'.format(filesafe(k), v) for k,v in self.config['method_opts'].items()])
+        return
+
+    # TODO: add dgp-specific, method-specific parameters and metrics
+    def experiment(self, instance_params, seed):
+        ''' Runs an experiment on a single randomly generated instance and sample and returns
+        the parameter estimates for each method and the evaluated metrics for each method
+        '''
+        np.random.seed(seed)
+
+        param_estimates = {}
+        true_params = {}
+        for dgp_name, dgp_fn in self.config['dgps'].items():
+            data, true_param = dgp_fn(self.config['dgp_opts'][dgp_name], instance_params[dgp_name], seed)
+            true_params[dgp_name] = true_param
+            param_estimates[dgp_name] = {}
+
+            for method_name, method in self.config['methods'].items():
+                param_estimates[dgp_name][method_name] = method(data, self.config['method_opts'][method_name], seed)
+
+        return param_estimates, true_params
+
+    def run(self):
+        ''' Runs multiple experiments in parallel on randomly generated instances and samples and returns
+        the parameter estimates for each method and the evaluated metrics for each method across all
+        experiments
+        '''
+        random_seed = self.config['mc_opts']['seed']
+
+        if not os.path.exists(self.config['target_dir']):
+            os.makedirs(self.config['target_dir'])
+
+        instance_params = {}
+        for dgp_name in self.config['dgps']:
+            instance_params[dgp_name] = self.config['dgp_instance_fns'][dgp_name](self.config['dgp_opts'][dgp_name], random_seed)
+
+        # results_file = os.path.join(self.config['target_dir'], 'results_{}.jbl'.format(self.config['param_str']))
```

Comment:
Throughout this file there is lots of dead code; delete all of it unless it serves some valuable purpose, in which case add comments explaining why.